File: android/src/main/java/com/aptoslabs/arculus/csdk/WalletStoreDataPIN.kt
package com.aptoslabs.arculus.csdk

import com.sun.jna.Pointer

class WalletStoreDataPIN(wallet: Pointer, private val pin: String) : CSDKAPICall<Unit>(wallet) {
  override suspend fun request(): Array<ByteArray> {
    val pinBytes = pin.encodeToByteArray()

    val len = SizeTByReference()

    val pointer = CSDK.WalletStoreDataPINRequest(wallet, pinBytes, pinBytes.size, len)

    if (pointer == Pointer.NULL) throw CSDKAPICallRequestCreationException()

    return arrayOf(pointer.getByteArray(0, len.toIntChecked()))
  }

  override suspend fun response(bytes: ByteArray) {

    val rc = CSDK.WalletStoreDataPINResponse(wallet, bytes, bytes.size)

    validateWalletResponse(rc)
  }
}

File: ios/Arculus/CSDK/WalletStoreDataPIN.swift
import CSDK

final class WalletStoreDataPIN: CSDKAPICall<Void> {
    private let pin: String
    
    init(wallet: OpaquePointer, pin: String) {
        self.pin = pin
        
        super.init(wallet: wallet)
    }
    
    override func request() async throws -> [Data] {
        var len: size_t = 0
        
        guard let bytes = WalletStoreDataPINRequest(wallet, pin, pin.count, &len) else {
            throw CSDKAPICallError(.requestCreationFailed("store PIN"))
        }
        
        return [Data(bytes: bytes, count: len)]
    }
    
    override func response(bytes: [UInt8]) async throws -> ResponseType {
        let result = WalletStoreDataPINResponse(wallet, bytes, bytes.count)
        
        try CSDKAPICallError.validateWalletResponseCode(rc: result)
    }
}
